src/Storages/MergeTree/MergeTreeMarksLoader.cpp

```diff
@@ -33,12 +43,49 @@ MergeTreeMarksLoader::MergeTreeMarksLoader(
     , columns_in_mark(columns_in_mark_)
     , read_settings(read_settings_)
 {
+    if (read_settings_.load_marks_asynchronously)
+    {
+        future = loadMarksAsync();
+    }
+}
+
+MergeTreeMarksLoader::~MergeTreeMarksLoader()
+{
+    if (future.valid())
```

Comment:
Interesting to check, at what moment in time, the future becomes invalid. Looks like some race condition is still possible.